### PR TITLE
dbt-materialize: exclude subsources from introspection

### LINKED ISSUE #20483: DROP SOURCE can no longer be used to drop Postgres subsources, breaks dbt
### What version of Materialize are you using?

 v0.61.0-dev (dcdb2bdab)

### What is the issue?

dbt can no longer drop postgres subsources when it wants to recreate them:

```
13:37:41  Database Error in model supplier (models/loadgen/supplier.sql)
13:37:41    SOURCE "supplier" is a subsource and must be dropped with ALTER SOURCE...DROP SUBSOURCE
13:37:41    HINT:  Use ALTER SOURCE tpch DROP SUBSOURCE supplier
```

A test can be made available if needed.

@sploiselle FYI

### PR DESCRIPTION
**Pull request by morsapaes**
Fixes #20483.

The issue was in the `materialize__list_relations_without_caching` macro, which listed subsources and progress subsources as `sources`. For multi-output sources, this meant that dbt would try to run a `DROP...CASCADE` statement for the source first, and then also for the subsources, which breaks. For future reference, you can get more detailed logging using `dbt --debug run`, @philip-stoev!

```sql
00:52:14  Began compiling node model.mz_get_started.accounts
00:52:14  Writing injected SQL for node "model.mz_get_started.accounts"
00:52:14  Timing info for model.mz_get_started.accounts (compile): 02:52:14.055838 => 02:52:14.061600
00:52:14  Began executing node model.mz_get_started.accounts
00:52:14  Using materialize connection "model.mz_get_started.accounts"
00:52:14  On model.mz_get_started.accounts: /* {"app": "dbt", "dbt_version": "1.5.2", "profile_name": "mz_get_started", "target_name": "dev", "node_id": "model.mz_get_started.accounts"} */

      drop source if exists "materialize"."qck"."accounts" cascade

00:52:14  Opening a new connection, currently in state closed
00:52:14  Materialize adapter: Enabling auto_route_introspection_queries
00:52:14  Postgres adapter: Postgres error: SOURCE "qck.accounts" is a subsource and must be dropped with ALTER SOURCE...DROP SUBSOURCE
HINT:  Use ALTER SOURCE qck.auction_house DROP SUBSOURCE qck.accounts

00:52:14  Postgres adapter: Error running SQL: macro drop_relation
00:52:14  Postgres adapter: Rolling back transaction.
00:52:14  Timing info for model.mz_get_started.accounts (execute): 02:52:14.062779 => 02:52:14.464442
00:52:14  On model.mz_get_started.accounts: Close
00:52:14  Database Error in model accounts (models/staging/accounts.sql)
  SOURCE "qck.accounts" is a subsource and must be dropped with ALTER SOURCE...DROP SUBSOURCE
  HINT:  Use ALTER SOURCE qck.auction_house DROP SUBSOURCE qck.accounts
```

Taking the chance to add very minimal quality-of-life improvements to `dbt docs`: populate `owner` in the dbt documentation, now that we have RBAC wired up, and use the `type` from `mz_sources` instead of `mz_objects` to correctly report subsources. The docs for the parent source object don't expose the object type, so we can live with it not being `source`, and rather one of `postgres`, `kafka` or `load-generator`
.
Screenshot 2023-07-13 at 03 50 37

### CODE REVIEW

File: misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql [hunk 1]
```diff
@@ -96,11 +96,17 @@
...
+        end as type
     from mz_objects o
+    left join mz_sources so on o.id = so.id
     join mz_schemas s on o.schema_id = s.id and s.name = '{{ schema_relation.schema }}'
     join mz_databases d on s.database_id = d.id and d.name = '{{ schema_relation.database }}'
-    where type in ('table', 'source', 'view', 'materialized-view', 'index', 'sink')
+    where o.type in ('table', 'source', 'view', 'materialized-view', 'index', 'sink')
+      --Exclude subsources and progress subsources, which aren't relevant in this
+      --context and can bork the adapter (see #20483)
+      and coalesce(so.type, '') not in ('subsource', 'progress')
```
Comment (philip-stoev) [inline misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:109]: strictly speaking no `coalesce` is needed here, `so.type not in ...` will work just as well.

**Review by philip-stoev — APPROVED** [review]
The patch worked for my case -- it is now possible to run `dbt run` twice without getting the error on the second try.

File: misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql (hunk 1 above)
Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:109]: I was surprised that...it doesn't.

Comment (philip-stoev) [reply misc/dbt-materialize/dbt/include/materialize/macros/adapters.sql:109]: Yes you are right, `NULL NOT IN ('something')` is NULL, that is `FALSE` , and you want it to be `TRUE` in order to match indexes and materialized views that have no type. My bad.

**Comment by morsapaes** [comment]
Added a test for subsources that also helps guarantee no other custom materializations error on consecutive dbt runs (which is where most issues arise).

**Pull request merged**